File: python_backend/game_logic/ranges/top/get_final_ranges.py
from game_logic.restriction.restriction import restriction
from game_logic.piece_type.get_piece_type import get_piece_type


def get_final_ranges(init_ranges, pins, threat_area, final_ranges, mt_restriction):
    """the final range is the intersection of the initial range, pin, threat area, and multi-threat restriction
    :param init_ranges:
    :param pins:
    :param threat_area:
    :param final_ranges:
    :param mt_restriction:
    :return:
    """
    for id_ in final_ranges.keys():
        if get_piece_type(id_) == 'K':
            final_ranges[id_] = init_ranges[id_]
        else:
            final_ranges[id_] = list(set(init_ranges[id_]) & restriction(pins[id_]) & restriction(threat_area) & restriction(mt_restriction[id_]))
    return final_ranges


if __name__ == "__main__":
    pass  # TODO: implement test for the final ranges

File: app/src/game_logic/council_logic/getFinalRanges.js
import { restriction } from "../restriction/restriction";
import { getPieceType } from "../pieceType/getPieceType";
import { mapListXyToRf } from "../coordType/mapListXyToRf";
import { mapDictListXyToRf } from "../coordType/mapDictListXyToRf";
import { intersection } from "../helpers/setOps";

// initRanges, pins, threatArea, finalRanges, mtRestrict
export function getFinalRanges(initRanges, threatArea, finalRanges) {
    /**the final range is the intersection of the initial range, pin, threat area, &&  multi-threat restriction
    :param initRanges:
    :param pins:
    :param threatArea:
    :param finalRanges:
    :param mtRestriction:
    :return:
    */

    initRanges = mapDictListXyToRf(initRanges);
    threatArea = mapListXyToRf(threatArea);

    let setA, setB, setC;

    for (const id of Object.keys(finalRanges)) {
        if (getPieceType(id) === "K") {
            finalRanges[id] = initRanges[id];
        } else {
            setA = new Set(initRanges[id]);
            setB = restriction(threatArea);
            setC = intersection(setA, setB);
            finalRanges[id] = Array.from(setC);
        }

        finalRanges[id] = Array.from(finalRanges[id]);
    }

    return finalRanges;
}

// module.exports = getFinalRanges;

/* for node.js
if (require.main === module) {
    import {printBoard} from "../printers/printBoard";
    import {sampleBoardDicts} from "../testObjects/sampleBoardDicts";
    import {getPieceLoc} from "../helpers/getPieceLoc";

    var board, initPieceDict, initRanges, threatArea, finalRanges;
}
*/
